examples/deepseek_v3/model_pytorch.py

```diff
@@ -0,0 +1,1854 @@
+        (self.scaling_factor * seq_len / self.max_position_embeddings)
+        - (self.scaling_factor - 1)
+      ) ** (self.dim / (self.dim - 2))
+      inv_freq = 1.0 / (
+        base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim)
+      )
+      self.register_buffer('inv_freq', inv_freq, persistent=False)
+
+    t = torch.arange(
+      self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype
+    )
+
+    freqs = torch.outer(t, self.inv_freq)
+    # Different from paper, but it uses a different permutation in order to obtain the same calculation
+    emb = torch.cat((freqs, freqs), dim=-1)
+    self.register_buffer('cos_cached', emb.cos().to(dtype), persistent=False)
+    self.register_buffer('sin_cached', emb.sin().to(dtype), persistent=False)
+
+
+# Inverse dim formula to find dim based on number of rotations
+def yarn_find_correction_dim(
+  num_rotations, dim, base=10000, max_position_embeddings=2048
+):
+  return (
+    dim * math.log(max_position_embeddings / (num_rotations * 2 * math.pi))
+  ) / (2 * math.log(base))
+
+
+# Find dim range bounds based on rotations
+def yarn_find_correction_range(
+  low_rot, high_rot, dim, base=10000, max_position_embeddings=2048
+):
+  low = math.floor(
+    yarn_find_correction_dim(low_rot, dim, base, max_position_embeddings)
+  )
+  high = math.ceil(
+    yarn_find_correction_dim(high_rot, dim, base, max_position_embeddings)
+  )
+  return max(low, 0), min(high, dim - 1)  # Clamp values just in case
+
+
+def yarn_get_mscale(scale=1, mscale=1):
+  if scale <= 1:
+    return 1.0
+  return 0.1 * mscale * math.log(scale) + 1.0
+
+
+def yarn_linear_ramp_mask(min, max, dim):
+  if min == max:
+    max += 0.001  # Prevent singularity
+
+  linear_func = (torch.arange(dim, dtype=torch.float32) - min) / (max - min)
+  ramp_func = torch.clamp(linear_func, 0, 1)
+  return ramp_func
+
+
+class DeepseekV3YarnRotaryEmbedding(DeepseekV3RotaryEmbedding):
+  def __init__(
+    self,
+    dim,
+    max_position_embeddings=2048,
+    base=10000,
+    device=None,
+    scaling_factor=1.0,
+    original_max_position_embeddings=4096,
+    beta_fast=32,
+    beta_slow=1,
+    mscale=1,
+    mscale_all_dim=0,
+  ):
+    self.scaling_factor = scaling_factor
+    self.original_max_position_embeddings = original_max_position_embeddings
+    self.beta_fast = beta_fast
+    self.beta_slow = beta_slow
+    self.mscale = mscale
+    self.mscale_all_dim = mscale_all_dim
+    super().__init__(dim, max_position_embeddings, base, device)
+
+  def _set_cos_sin_cache(self, seq_len, device, dtype):
+    self.max_seq_len_cached = seq_len
+    dim = self.dim
+
+    freq_extra = 1.0 / (
+      self.base
+      ** (torch.arange(0, dim, 2, dtype=torch.float32, device=device) / dim)
+    )
+    freq_inter = 1.0 / (
+      self.scaling_factor
+      * self.base
+      ** (torch.arange(0, dim, 2, dtype=torch.float32, device=device) / dim)
+    )
+
+    low, high = yarn_find_correction_range(
```

Comment:
Seams Huggingface implementation is different from the reference implementation, e.g.

```
    if seqlen > args.original_seq_len:
        low, high = find_correction_range(beta_fast, beta_slow, dim, base, args.original_seq_len)
        smooth = 1 - linear_ramp_factor(low, high, dim // 2)
        freqs = freqs / factor * (1 - smooth) + freqs * smooth

```

https://github.com/deepseek-ai/DeepSeek-V3/blob/2f7b80eecebf3d1c84da5a0d465f6639ea175012/inference/model.py#L364